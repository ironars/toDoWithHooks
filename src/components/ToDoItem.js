import React from 'react'
import { useDispatch } from 'react-redux'
import { completeToDo } from '../actions'

function ToDoItem({ todoId, todo }) {
  const dispatch = useDispatch()

  const handleComplete = (todoItem) => {
    dispatch(completeToDo(todoItem, !todo.isCompleted))
  }

  return (
    <li className="t-item">
      <input
        onChange={() => handleComplete(todoId)}
        type="checkbox"
        id={todoId}
        name={todoId}
        value={todo.isCompleted}
        checked={todo.isCompleted}
      ></input>
      <label htmlFor={todoId} className={todo.isCompleted ? 'completed' : ''}>
        {todo.title}
      </label>
    </li>
  )
}

export default ToDoItem
