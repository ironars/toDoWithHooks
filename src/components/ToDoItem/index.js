import React from 'react'
import { string, bool, func } from 'prop-types'

function ToDoItem({ task, isCompleted, clickHandler }) {
  if (!task) return null
  const cx = isCompleted ? 'completed' : ''
  return (
    <li className="t-item">
      <input
        onChange={clickHandler}
        type="checkbox"
        data-testid="todoCheck"
        name={task}
        value={isCompleted}
        checked={isCompleted}
      />
      <label htmlFor={task} className={cx}>
        {task}
      </label>
    </li>
  )
}

ToDoItem.propTypes = {
  task: string.isRequired,
  isCompleted: bool,
  clickHandler: func
}

ToDoItem.defaultProps = {
  isCompleted: false,
  clickHandler: () => {}
}

export default ToDoItem
