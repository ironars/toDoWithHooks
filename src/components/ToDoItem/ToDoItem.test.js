import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
// import jest from 'jest'

import ToDoItem from './index'

describe('ToDoItem', () => {
  it('should render an incomplete task', () => {
    const { getByText } = render(<ToDoItem task={'Meeting'} />)
    const node = getByText('Meeting')
    expect(node.textContent).toBe('Meeting')
  })

  it('should not render if a task is empty', () => {
    const { container } = render(<ToDoItem task={''} />)
    const node = container.firstChild
    expect(node).not.toBeInTheDocument()
    expect(node).toBe(null)
  })

  xit('should not render if a task is undefined', () => {
    const { container } = render(<ToDoItem />) //task is undefined here
    const node = container.firstChild
    expect(node).not.toBeInTheDocument()
    expect(node).toBe(null)
  })

  it('should render and match snapshot', () => {
    const { container } = render(<ToDoItem task="Meeting" />)

    expect(container).toMatchSnapshot()
  })

  it('should show the value of complete task', () => {
    const { getByTestId } = render(<ToDoItem task="Meeting" isCompleted />)
    const node = getByTestId('todoCheck')
    expect(node.checked).toBe(true)
    expect(node.checked).not.toBe(false)
  })

  it('should call the handleClick function', () => {
    const mockFn = jest.fn(() => {})
    const { getByTestId } = render(<ToDoItem task="Meeting" clickHandler={mockFn} />)
    const node = getByTestId('todoCheck')
    fireEvent.click(node)
    expect(mockFn).toHaveBeenCalled()
    expect(mockFn).toHaveBeenCalledTimes(1)
  })
})
