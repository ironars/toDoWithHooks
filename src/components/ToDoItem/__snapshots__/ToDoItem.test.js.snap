// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoItem should render and match snapshot 1`] = `
<div>
  <li
    class="t-item"
  >
    <input
      data-testid="todoCheck"
      name="Meeting"
      type="checkbox"
      value="false"
    />
    <label
      class=""
      for="Meeting"
    >
      Meeting
    </label>
  </li>
</div>
`;
